File: Gfg/Huffman Encoding.cpp
#include<vector>
using namespace std;

struct MinHeapNode {
    char symbol;
    int freq;
    MinHeapNode *left, *right;
    MinHeapNode(char symbol, int freq)
    {
        this->left = this->right = nullptr;
        this->symbol = symbol;
        this->freq = freq;
    }
};

struct compare {
    bool operator()(MinHeapNode *left, MinHeapNode *right)
    {
        return (left->freq > right->freq);    
    }
};

void printHuffCodes(MinHeapNode *root, string str)
{
    if (root == nullptr)
        return;
    
    if (root->symbol != '#') {
        //cout << root->symbol << " " << str << " ";
        cout << str << " ";
    }
    if (root->left) printHuffCodes(root->left, str + '0');
    if (root->right) printHuffCodes(root->right, str + '1');
}

void huffmanEncode(string str, int *freq, int n)
{
    priority_queue<MinHeapNode*, vector<MinHeapNode*>, compare> minHeap;
    
    for (int i = 0; i < n; i++) {
        minHeap.push(new MinHeapNode(str[i], freq[i]));
        //cout << str[i] << " " << freq[i] << " ";
    }
    
    while (minHeap.size() > 1) {
        MinHeapNode *left = minHeap.top();
        minHeap.pop();
        
        MinHeapNode *right = minHeap.top();
        minHeap.pop();
        
        MinHeapNode *cur = new MinHeapNode('#', left->freq + right->freq);
        cur->left = left;
        cur->right = right;
        //cout << cur->symbol << " " << cur->freq << " ";
        //cout << cur->left->symbol << " " << cur->right->freq << " ";
        //cout << cur->left->symbol << " " << cur->right->freq << " ";
        minHeap.push(cur);
    }
    
    printHuffCodes(minHeap.top(), "");
}

int main()
{
    // Note that size of arr[] is considered 100 according to
    // the constraints mentioned in problem statement.
    int freq[100], x, t, n;
    string str;
 
    // Input the number of test cases you want to run
    cin >> t;
 
    // One by one run for all input test cases
    while (t--)
    {
        cin >> str;
        //cout << str << " " << str.size() << endl;
        // Input the array
        for (int i = 0; i < str.size(); i++) {
            cin >> freq[i];
        }

        // Compute and print result
        huffmanEncode(str, freq, str.size());
        cout << endl;
    }
    return 0;
}

File: Gfg/Huffman Encoding.java
import java.lang.*;
import java.io.*;
class GFG
 {
	public static void main (String[] args)
	 {
	 //code
	 Scanner s=new Scanner(System.in);
	 int T=s.nextInt();
	 for(int i=0;i<T;i++)
	 {
	     String str=s.next();
	     int[] arr=new int[str.length()];
	     for(int j=0;j<arr.length;j++)
	     {
	         arr[j]=s.nextInt();
	     }
	     encode(str,arr);
	     System.out.println();
	 }
	 }
	 
	 public static void encode(String str,int[] arr)
	{
		        PriorityQueue<Node> queue=new PriorityQueue<>(new Comparator<Node>()
				{
				@Override
				public int compare(Node n1,Node n2)
				{
				    if(n1.freq==n2.freq)
				    return 1;
					return n1.freq-n2.freq;
				}
				}
		);
		for(int i=0;i<str.length();i++)
		{
			Node node=new Node(""+str.charAt(i),arr[i],null,null);
			queue.add(node);
		}
		while(queue.size()>1)
		{
			Node node1=queue.remove();
			Node node2=queue.remove();
			Node node=new Node(node1.name+node2.name,node1.freq+node2.freq,node1,node2);
			queue.add(node);
		}
		huffmanencode(queue.peek(),"");
	}
	
	public static void huffmanencode(Node root,String res)
	{
		if(root.left==null&&root.right==null)
		{
			System.out.print(res+" ");
			return;
		}
		huffmanencode(root.left,res+"0");
		huffmanencode(root.right,res+"1");
	}
	public static class Node
	{
		String name;
		int freq;
		Node left;
		Node right;
		
		Node(String name,int freq,Node left,Node right)
		{
			this.name=name;
			this.freq=freq;
			this.left=left;
			this.right=right;
		}
	}
}
